Add support for ForeFlight format

zyv opened the pull request:
# Support for ForeFlight format

Implements #40. Enjoy!

So, here comes the basic support for the ForeFlight format. I've tried to break it down into atomic commits so that it's easier for you to understand and review, but I haven't ensured that each one compiles separately, that would just be way too much work.

I started testing it by converting the books, which seemed too painful on the iPad, and found some minor problems that I fixed in follow-up commits instead of squashing everything, just to give you an idea of the issues. Maybe with more testing I'll find something else, but for now it seems to work pretty well.

Below are some points for consideration and perhaps discussion.

### New fields added to the models

I've added `description` to your metadata model, because I couldn't find a sensible way to map it.

I've also added `type` to `categories` for the reasons discussed in #40:

ForeFlight has 3 predefined categories, and your categories are what they call subcategories. So either we need a more nested structure, or we need to flatten it somehow. My idea was to flatten it by mapping ForeFlight categories to category types.

Which brings us to the next point:

### Support for category types

I made sure they are ignored by default. What is missing is:

1. Ability to select type when creating a new category
2. Ability to change the type of an existing category
3. Unknown category type is not displayed (on purpose)

So currently you have to start with a ForeFlight checklist to be able to create one. My understanding is that asking for more properties would require a proper modal dialog. Not sure what you think. Would you like to help with this?

### Possible future refactorings

I noticed that the list of supported formats and their properties are currently scattered all over the application. It would be nice to create an `Enum` to centralize this.

Unfortunately, TypeScript doesn't seem to have object-based enums like Python, but they can be type-safe emulated with some minimal trickery: https://stackoverflow.com/a/50224786/5132088 .

## Open tasks before merging

- [x] Fix new proto file style to use `snake_case`
- [x] Fix handling of `name` metadata field - fallback to upload filename
- [x] Rework `objectId` generation and add documentation
- [x] Remove extra FF `groupType' enum and use EFIS group categories
- [x] Pass `ForeFlightUtils` directly to templates
- [x] Implement UI to change EFIS group category
- [x] Improve metadata and item mapping
  - [x] Better strategy for `description` field
  - [x] Decide on remaining item types
  - [x] Merge notes coming one after another
  - [x] Decide on mapping categories for other formats
- [x] ~~Rework test data, currently ForeFlight format has its own mock~~
- [x] Update format notes on the welcome page

zyv commented:
Ha-ha, got bit by my own linter :) Fixed the CI!

rdamazio commented on src/model/formats/foreflight-reader.ts:
```diff
@@ -0,0 +1,118 @@
+  /**
+   * Unfortunately, all fields in ForeFlight are optional, so take first one filled out or default
+   */
+  private static getChecklistName(metadata: ForeFlightChecklistMetadata): string {
```
Nitpicking: this is the FILE name, not a checklist name (probably just a nomenclature difference).

In other formats we set this to just the uploaded filename (minus the extension) - should we keep that consistent?

rdamazio replied:
> Anyway, I added https://github.com/rdamazio/efis-editor/commit/e2599f9255845c8cbc621d9c8f69a26299e5920a with what I think is the least bad solution

Thanks, that LGTM.

rdamazio commented on src/model/formats/foreflight-utils.ts:
```diff
@@ -0,0 +1,93 @@
+    );
+  }
+
+  public static categoryToEFIS(category: string): ChecklistGroup_Category {
```
Doesn't proto support defining the field as the enum directly and parsing from the strings, instead of needing constants + this function?

zyv replied:
Unfortunately, not to my knowledge, at least not according to this issue: https://github.com/protocolbuffers/protobuf/issues/8483. The documentation that I've found says that enums must be integer-based (https://protobuf.dev/programming-guides/proto3/#enum) and the generated code for `ChecklistGroup_Category` doesn't even include a reverse mapping.

I could try to do some reflection magic instead of a switch for the conversion though, and I would have done it like 10 years ago, but these days I feel like the complexity and type erasure is not worth it for 3 elements...

Please let me know if you have a better solution, maybe I'm still missing something.

rdamazio replied:
That's weird, 'cause when I export the protos in JSON format ("raw format" in the UI), it gives me output like this:

```
            {
              "prompt": "Fuel required",
              "expectation": "COMPUTED",
              "type": "ITEM_CHALLENGE_RESPONSE"
            },
```

which it also parses back just fine - so it is using the string of the enum in the JSON.

I just downloaded your code and took a stab at this - here's an (incomplete) patch. The key thing is parsing the JSON with the proto class itself. The caveat that I'm not sure there's a workaround for (and why the patch is incomplete, I didn't have time so far to dig further) is that the parser seems to be case-sensitive for the enum names :/

```
diff --git a/src/model/formats/foreflight-format.spec.ts b/src/model/formats/foreflight-format.spec.ts
index 75f4e21530...d56764ec4f 100644
--- a/src/model/formats/foreflight-format.spec.ts
+++ b/src/model/formats/foreflight-format.spec.ts
@@ -1,5 +1,4 @@
-import { ChecklistGroup_Category } from '../../../gen/ts/checklist';
-import { ForeFlightCategory, ForeFlightUtils } from './foreflight-utils';
+import { ForeFlightUtils } from './foreflight-utils';
 import { EXPECTED_FOREFLIGHT_CONTENTS } from './test-data';
 import { ForeFlightReader } from './foreflight-reader';
 import { loadFile } from './test-utils';
@@ -22,13 +21,6 @@
       expect(decrypted).toBe(checklistData);
     });

-    it('should map ForeFlight categories to EFIS', async () => {
-      expect(ForeFlightUtils.categoryToEFIS(ForeFlightCategory.Normal)).toBe(ChecklistGroup_Category.NORMAL);
-      expect(() => {
-        ForeFlightUtils.categoryToEFIS('undefined');
-      }).toThrowError("ForeFlight: unknown category 'undefined'");
-    });
-
     it('should determine correct checklist file name', async () => {
       const mockFile = new File([], `bar.${ForeFlightUtils.FILE_EXTENSION}`);
       const mockContainer = ForeFlightChecklistContainer.create({ payload: { metadata: { name: 'foo' } } });
diff --git a/src/model/formats/foreflight-reader.ts b/src/model/formats/foreflight-reader.ts
index a839f2d0ca...aa21426542 100644
--- a/src/model/formats/foreflight-reader.ts
+++ b/src/model/formats/foreflight-reader.ts
@@ -19,7 +19,7 @@
 export class ForeFlightReader {
   public static async read(file: File): Promise<ChecklistFile> {
     const plaintext = await ForeFlightUtils.decrypt(await file.arrayBuffer());
-    const container: ForeFlightChecklistContainer = JSON.parse(plaintext);
+    const container: ForeFlightChecklistContainer = ForeFlightChecklistContainer.fromJsonString(plaintext);

     if (container.type !== ForeFlightUtils.CONTAINER_TYPE) {
       throw new ForeFlightFormatError(`unknown checklist type '${container.type}'`);
@@ -56,9 +57,8 @@
   }

   private static checklistGroupToEFIS(checklistGroup: ForeFlightChecklistGroup): ChecklistGroup[] {
-    const category = ForeFlightUtils.categoryToEFIS(checklistGroup.groupType);
     return checklistGroup.items
-      ? checklistGroup.items.map((item) => ForeFlightReader.checklistSubgroupToEFIS(category, item))
+      ? checklistGroup.items.map((item) => ForeFlightReader.checklistSubgroupToEFIS(checklistGroup.groupType, item))
       : []; // if the FF group (not subgroup!) is empty, drop it - we need a subgroup to represent a group
   }

diff --git a/src/model/formats/foreflight-utils.ts b/src/model/formats/foreflight-utils.ts
index 4cacacaf35...8e038cb288 100644
--- a/src/model/formats/foreflight-utils.ts
+++ b/src/model/formats/foreflight-utils.ts
@@ -1,14 +1,7 @@
-import { ChecklistGroup_Category } from '../../../gen/ts/checklist';
 import { v4 as uuid } from 'uuid';
 import { ForeFlightFormatError } from './foreflight-format';
 import { ForeFlightChecklistContainer } from '../../../gen/ts/foreflight';

-export enum ForeFlightCategory {
-  Normal = 'normal',
-  Abnormal = 'abnormal',
-  Emergency = 'emergency',
-}
-
 enum KeyUsage {
   Encrypt = 'encrypt',
   Decrypt = 'decrypt',
@@ -69,19 +62,6 @@
     );
   }

-  public static categoryToEFIS(category: string): ChecklistGroup_Category {
-    switch (category) {
-      case ForeFlightCategory.Normal:
-        return ChecklistGroup_Category.NORMAL;
-      case ForeFlightCategory.Abnormal:
-        return ChecklistGroup_Category.ABNORMAL;
-      case ForeFlightCategory.Emergency:
-        return ChecklistGroup_Category.EMERGENCY;
-      default:
-        throw new ForeFlightFormatError(`unknown category '${category}'`);
-    }
-  }
-
   public static async getObjectId(): Promise<string> {
     const hash = Array.from(
       new Uint8Array(await window.crypto.subtle.digest('SHA-256', ForeFlightUtils.encoder.encode(uuid()))),
diff --git a/src/model/formats/foreflight-writer.ts b/src/model/formats/foreflight-writer.ts
index 9bc199ce70...e2080079e8 100644
--- a/src/model/formats/foreflight-writer.ts
+++ b/src/model/formats/foreflight-writer.ts
@@ -1,11 +1,11 @@
-import { ChecklistFile, ChecklistGroup, ChecklistItem, ChecklistItem_Type } from '../../../gen/ts/checklist';
+import { ChecklistFile, ChecklistGroup, ChecklistGroup_Category, ChecklistItem, ChecklistItem_Type } from '../../../gen/ts/checklist';
 import {
   ForeFlightChecklistContainer,
   ForeFlightChecklistGroup,
   ForeFlightChecklistItem,
 } from '../../../gen/ts/foreflight';

-import { ForeFlightCategory, ForeFlightUtils } from './foreflight-utils';
+import { ForeFlightUtils } from './foreflight-utils';

 export class ForeFlightWriter {
   public static async write(file: ChecklistFile): Promise<Blob> {
@@ -32,14 +32,14 @@

   private static async checklistGroupsToFF(groupsEFIS: ChecklistGroup[]): Promise<ForeFlightChecklistGroup[]> {
     return Promise.all(
-      Object.values(ForeFlightCategory).map(
+      [ChecklistGroup_Category.NORMAL, ChecklistGroup_Category.ABNORMAL, ChecklistGroup_Category.EMERGENCY].map(
         async (categoryFF): Promise<ForeFlightChecklistGroup> => ({
           objectId: await ForeFlightUtils.getObjectId(),
           groupType: categoryFF,
           items: await Promise.all(
             groupsEFIS
               .filter(
-                (checklistGroupEFIS) => checklistGroupEFIS.category === ForeFlightUtils.categoryToEFIS(categoryFF),
+                (checklistGroupEFIS) => checklistGroupEFIS.category === categoryFF,
               )
               .map(async (checklistGroupEFIS) => ({
                 objectId: await ForeFlightUtils.getObjectId(),
diff --git a/src/model/proto/foreflight.proto b/src/model/proto/foreflight.proto
index 2ea5ca18ac...dc1f79f880 100644
--- a/src/model/proto/foreflight.proto
+++ b/src/model/proto/foreflight.proto
@@ -4,6 +4,8 @@

 option optimize_for = CODE_SIZE;

+import "checklist.proto";
+
 message ForeFlightChecklistContainer {
   string type = 1; // Always "checklist"
   ForeFlightChecklistPayload payload = 2;
@@ -25,7 +27,7 @@

 message ForeFlightChecklistGroup {
   string object_id = 1;
-  string group_type = 2;
+  efis.ChecklistGroup.Category group_type = 2;
   repeated ForeFlightChecklistSubgroup items = 3; // This key might be missing altogether
 }
```

If you see an easy solution for the case-sensitivity, great, if not, we can just leave a TODO or something to simplify it later.

rdamazio replied:
Additional patch that does make it work:

```
diff --git a/src/model/formats/foreflight-writer.ts b/src/model/formats/foreflight-writer.ts
index e2080079e8...4340c9ea8f 100644
--- a/src/model/formats/foreflight-writer.ts
+++ b/src/model/formats/foreflight-writer.ts
@@ -32,7 +32,7 @@
 
   private static async checklistGroupsToFF(groupsEFIS: ChecklistGroup[]): Promise<ForeFlightChecklistGroup[]> {
     return Promise.all(
-      [ChecklistGroup_Category.NORMAL, ChecklistGroup_Category.ABNORMAL, ChecklistGroup_Category.EMERGENCY].map(
+      [ChecklistGroup_Category.normal, ChecklistGroup_Category.abnormal, ChecklistGroup_Category.emergency].map(
         async (categoryFF): Promise<ForeFlightChecklistGroup> => ({
           objectId: await ForeFlightUtils.getObjectId(),
           groupType: categoryFF,
diff --git a/src/model/formats/test-data.ts b/src/model/formats/test-data.ts
index 43a91739ce...d37a61ffb8 100644
--- a/src/model/formats/test-data.ts
+++ b/src/model/formats/test-data.ts
@@ -87,14 +87,14 @@
 
 export const EXPECTED_FOREFLIGHT_CONTENTS = {
   groups: [
-    { category: ChecklistGroup_Category.NORMAL, title: 'Empty subgroup', checklists: [] },
+    { category: ChecklistGroup_Category.normal, title: 'Empty subgroup', checklists: [] },
     {
-      category: ChecklistGroup_Category.NORMAL,
+      category: ChecklistGroup_Category.normal,
       title: 'Subgroup with empty checklist',
       checklists: [{ title: 'Empty checklist', items: [] }],
     },
     {
-      category: ChecklistGroup_Category.NORMAL,
+      category: ChecklistGroup_Category.normal,
       title: 'Subgroup',
       checklists: [
         {
@@ -162,7 +162,7 @@
         },
       ],
     },
-    { category: ChecklistGroup_Category.ABNORMAL, title: 'Test empty abnormal subgroup', checklists: [] },
+    { category: ChecklistGroup_Category.abnormal, title: 'Test empty abnormal subgroup', checklists: [] },
   ],
   metadata: {
     name: 'Checklist name',
diff --git a/src/model/proto/checklist.proto b/src/model/proto/checklist.proto
index 6f01f0c6d1...d2ae74adff 100644
--- a/src/model/proto/checklist.proto
+++ b/src/model/proto/checklist.proto
@@ -26,11 +26,13 @@
   string title = 1;
   repeated Checklist checklists = 2;
 
+  // TODO: Make these uppercase when ts-protobuf supports customizing
+  // JSON output to match Foreflight's expectations.
   enum Category {
     UNKNOWN = 0;
-    NORMAL = 1;
-    ABNORMAL = 2;
-    EMERGENCY = 3;
+    normal = 1;
+    abnormal = 2;
+    emergency = 3;
   }
 
   Category category = 3;
```

zyv replied:
Thanks for the patch, this clears up a lot of things for me! I was missing one non-technical and two technical items:

1. I wasn't sure if you would accept a FF-based category enum in your model, so I decided to decouple it first.
2. I missed `fromJsonString` in the protobuf documentation.
3. Ditto for the lowercase/string enum representation trick.

I have now committed a new version in a5de561 that hopefully addresses your concerns:

1. Mapping is removed, protos are coupled.
2. "Unknown" member is removed, because if there is no mapping, I see no value in it.


I've also tried to be more consistent with the teminology.

If you don't like the magic of enum iteration, I can change it to using hardcoded keys. I'm always afraid to do things like that, as it's easier to change things in one place and forget it in another.

I left your comment about customizing the output, but I doubt that'll ever happen:

https://github.com/golang/protobuf/issues/613

The last comment is pretty clear:

> If your service really needs to produce enum values in lowercase, then perhaps define them in lowercase in the proto file.

I'd rather remove the TODO if you don't mind.

rdamazio replied:
Thanks, what you pushed LGTM.

> I'd rather remove the TODO if you don't mind.

Yes, I'm fine with that.

zyv replied:
Removed TODO in 8da2814.

rdamazio commented on src/model/proto/foreflight.proto:
```diff
@@ -0,0 +1,52 @@
+// This would have been better modelled by inheritance, or better still using oneOf.
+// Unfortunately, inheritance is not supported and oneOf cannot be used with repeated fields.
+message ForeFlightChecklistItem {
+  string objectId = 1;
```
Proto field naming is usually object_id, group_type, etc. (and the generated code and the JSON field names should still be the same as you're using now)

https://protobuf.dev/programming-guides/proto3/#json

zyv replied:
Thanks for the hint! I missed the part about automagic mapping :( Fixed in 6425adc.

rdamazio commented on src/model/formats/foreflight-utils.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+  }
+
+  public static async getObjectId(): Promise<string> {
```
Can you add a comment explaining what the Object ID is? Is it any arbitrary identifier for each entry/checklist/etc? Is is a specific format?

zyv replied:
See below ☺️

zyv replied:
Added explanations from this PR to the proto file in 7580f05.

rdamazio commented on src/model/formats/foreflight-utils.ts:
```diff
@@ -0,0 +1,93 @@
+
+  public static async getObjectId(): Promise<string> {
+    const hash = Array.from(
+      new Uint8Array(await window.crypto.subtle.digest('SHA-256', ForeFlightUtils.encoder.encode(uuid()))),
```
If the Object ID is NOT in a specific format, any reason not to just take the UUID (or part of it) and use it directly without hashing? the UUID is already guaranteed unique.

A consequence would be that you can get rid of most async/Promise handling in the writing code and make it simpler.

zyv replied:
That was a majestic brain fart, thanks for speaking up! Fixed in dd1acba.

Obviously, I don't have any official information on what `objectId` is and how it is used, but my guess is that it is used in some way to synchronize checklists between devices. I made the following observations:

1. It always seems to be 32 characters long.
2. It always seems to contain characters from `[a-f0-9]`.
3. Objects with the same properties have different `objectId`.

It looked very much like 128 bits of "something" encoded as hex, but that "something" somehow doesn't seem to be a hash of the object. So, to try to keep our files as indistinguishable from the original files as possible to hopefully avoid possible breakage, my spontaneous idea was to take UUID V4... but wait, they're 36 characters long, so I'll just hash them with SHA and cut them down to 32 characters, hopefully not increasing the chance of collisions too much.

Except, you know...

1. UUIDs actually carry exactly 128 bits, which is what I needed in the first place...
2. ...and most embarrassingly, upon closer inspection, the `objectId`s in ForeFlight-generated files are all valid UUID V4 with the dashes removed :-)

All I can say is that I got too carried away in my hacking spree.

rdamazio commented on src/model/formats/foreflight-writer.ts:
```diff
@@ -0,0 +1,97 @@
+          if (
+            lastItemFF &&
+            lastItemFF.type !== ForeFlightUtils.ITEM_HEADER &&
+            itemEFIS.indent === ForeFlightUtils.NOTE_INDENT
```
Shouldn't the note indent be 1 more than the item's indent, rather than always 1?
In any case, I don't think we should drop the note on the floor if it's not the right indent.

zyv replied:
Good point! Changed accordingly in ac46d25.

rdamazio commented on src/model/formats/test-foreflight.fmd:
Just checking - was this a file that Foreflight generated, or this app? I'd prefer the former

zyv replied:
Yes, it is a genuine ForeFlight generated file, exactly for the reasons you ask ;) If I need to make changes to the test data, I will upload the checklist to ForeFlight and have it regenerate the export file.

rdamazio commented on src/model/formats/test-data.ts:
```diff
@@ -84,3 +84,86 @@ export const EXPECTED_CONTENTS: ChecklistFile = {
   },
 };
+
+export const EXPECTED_FOREFLIGHT_CONTENTS = {
```
Is this because we don't (yet?) support features like indenting/centering in Foreflight files? If so, please add a TODO to add that support later (it's only natively supported on Garmin, but we can play with ways to represent the same elsewhere, as in other formats).

zyv replied:
Initially, I started with a separate mock because my goal was to achieve a roundtrip, and you have to start with something manageable. I generally agree that it is desirable to have the same mock for all formats.

However, now that I've finished the mapping, I'm more and more inclined to try to push you to leave it as is. The reason is that the ForeFlight format is not line or field based, and the interesting test cases are not field lengths or setting some property, but rather all kinds of missing pieces and dangerous item order. That's how I developed my mock - the ForeFlight file has all the missing things that can go missing, and I'm also testing item order and note collapsing.

We can try to merge the mocks, but the ForeFlight format doesn't support it, and I don't see a good way to introduce support for the following features:

1. Item indentation (except for attaching text items to headings and/or challenges as notes)
2. Centering of items
3. Metadata

I've tried to make the test data in 666ec05 a bit more grokable, but I'm not sure if it helps my cause.

If you still want to merge them, we can try after the metadata story is behind us, but I'd like to ask that I be able to move my groups over in addition to the existing ones, so that the edge cases I mentioned above are all properly tested and I have to do some magic to ignore unsupported stuff for ForeFlight. I'd need your help to regenerate the other files though, as I don't have access to other avionics.

rdamazio replied:
> I don't have access to other avionics

I only have access to Garmin, btw.

> more inclined to try to push you to leave it as is

I'm ok with this. We can also play more with formatting afterwards and see if we come up with something we like, no need to get it all right on this PR.

rdamazio commented on src/app/checklists/checklists.component.ts:
```diff
@@ -53,6 +55,9 @@ export class ChecklistsComponent implements OnInit {
   showFilePicker = false;
   showFileUpload = false;
 
+  protected readonly FOREFLIGHT_EXTENSION = ForeFlightUtils.FILE_EXTENSION;
```
Optional: I *think* you can reference other packages from the template, if you want to avoid this.

zyv replied:
I tried to pass only the variables that should be used instead of the whole class to control the use of its contents. I changed it in 416cdd7 if you like it better.

rdamazio commented on src/app/welcome/welcome.component.html:
```diff
@@ -66,4 +67,21 @@ <h1>Welcome to the EFIS editor!</h1>
+          types
+        </li>
+      </ul>
+      Other items will be silently ignored when exporting the checklist file!
```
Can we instead convert them to Challenge type? It's probably better to lose formatting than to lose the whole item.

zyv replied:
I implemented full mapping in ac46d25 and changed the note in e8d3999. I think it's still useful to point out that different formats can represent checklists in possibly unexpected ways.

rdamazio commented:
This is amazing, thank you so much for the PR!!
A few comments, but nothing major.

rdamazio commented on src/model/formats/foreflight-writer.ts:
```diff
@@ -0,0 +1,97 @@
+      Object.values(ForeFlightCategory).map(
+        async (categoryFF): Promise<ForeFlightChecklistGroup> => ({
+          objectId: await ForeFlightUtils.getObjectId(),
+          groupType: categoryFF,
```
Nitpicking: call it either category or group type uniformly

rdamazio replied:
With your latest change for object IDs, you should be able to remove *all* async and aawait from these functions (except write() probably)

zyv replied:
Thanks, I naively expected the linter to warn me about this, so I missed it. Makes everything much easier! Fixed in 99b41d2.

rdamazio commented:
Hmmm something else is not right - I tried to export my plane's checklists as FMD with this code, and it gave me a 560-byte file which was empty :/

Original file: [N425RP.json](https://github.com/user-attachments/files/16644684/N425RP.json)
Written contents:
```
{
  "type": "checklist",
  "payload": {
    "objectId": "42eb846306a4322f1009a839e52c89a2",
    "schemaVersion": "1.0",
    "metadata": {
      "name": "N425RP",
      "tailNumber": "N425RP"
    },
    "groups": [
      {
        "objectId": "433d476f7f529883ea5f472a462232ae",
        "groupType": "normal"
      },
      {
        "objectId": "1abc3aa6d3370b49caa08f3cba183001",
        "groupType": "abnormal"
      },
      {
        "objectId": "d04a8f7cab143b404aa5209beae25749",
        "groupType": "emergency"
      }
    ]
  }
}
```

zyv commented on src/model/proto/checklist.proto:
```diff
@@ -13,6 +13,8 @@ message ChecklistFileMetadata {
   string manufacturer_info = 6;
   string copyright_info = 7;
+
+  optional string description = 8;
```
I gave a partial explanation for this in my previous comment. ForeFlight has only 3 metadata fields, all optional, that do not match the fields you have for a perfect 1:1 translation:

* Filename can sort of match "name", with the caveats in the other comment.
* Tail number can be perfectly matched 1:1
* Description doesn't obviously match any of your fields

My priority was to make sure that when I load an FF checklist and then re-export it, the metadata is as unaffected as possible. This kind of requires setting the "Name" field in the original ForeFlight checklist, but I think it's an acceptable compromise.

As for the other fields, I mapped "Tail number" 1:1 and introduced "Description" to allow 1:1 mapping. All other fields are simply ignored, which is of course suboptimal, because the information entered in them will be completely lost, unless the checklist is additionally exported in the native format.

Now there are several ways to go about this in line with your suggestions:

1. Accept my solution of adding a generic description field (and maybe implement it for other formats?)
2. Map "description" 1:1 to an existing field like "manufacturer information" instead of adding a new field (and possibly implement aggregation of all other fields into this field - the downside is that aggregated data can't be broken down properly again).

I don't have a strong opinion on this, but I have no idea about other formats, and unfortunately I'm not sure how to get a feel for how they actually display. I don't have easy access to SkyView, and they don't seem to have a simulator, and I'm not sure if I can load checklists into the Garmin simulator. So I don't know if and how other formats can benefit from a generic "description" field.

Should I implement the second option (map to "manufacturer information"), and if so, with or without aggregation? I guess that this option with aggregation is the least bad in our context?

rdamazio replied:
Here's how I'd approach this - I'd add the description field, then when loading:
- If a line starts with a known prefix like "Manufacturer info: ", map it to that field
- All lines that are left go into the description field
- For other formats that are not field-restricted (i.e. not Garmin) I can also plug in the description field in the same way

When writing:
- Output the description field first
- For each field that's filled out, output it with a prefix

That lets us connect the full functionality to most formats.

I can send a commit doing this for other formats if you think that's reasonable, then you can just follow the same.

zyv replied:
I think I'm mostly done now, and your suggestion sounds reasonable. I'd appreciate it if you'd add a commit for other formats, so I can implement it for ForeFlight as well.

rdamazio replied:
Most of it is already there in text-based formats: https://github.com/rdamazio/efis-editor/blob/main/src/model/formats/text-reader.ts#L259
I can add the description part to that once we merge your changes.

zyv replied:
Alright, I implemented the two-way translation as discussed and started testing with the app, and that's when disaster struck.

What I didn't consider is that the `description` field does not allow multiline input! And even on a single line, there are only a few dozen characters you can fit before ForeFlight truncates it. Editing the long lines is also super hard.

On the other hand, by studying the manual, I discovered that the intended use is apparently what you call "make and model". So I just mapped it that way.

<img width="740" alt="Screenshot 2024-08-26 at 09 17 15" src="https://github.com/user-attachments/assets/1fd8e574-4c12-4f8e-aac3-64303d487b28">

This leaves us with the copyright and manufacturer information being lost on ForeFlight exports, but at least everything else works as expected. What do you think, can you live with this?

rdamazio replied:
Yes, this seems fine and idiomatic. We can revise if anyone complains.

zyv commented:
> Original file: N425RP.json

I can't believe it, but apparently it wasn't possible to import the native format before, so I implemented it in d6f7b58. I hope you don't mind.

> Hmmm something else is not right - I tried to export my plane's checklists as FMD with this code, and it gave me a 560-byte file which was empty :/

This was expected with my original implementation which decoupled EFIS categories from ForeFlight groups. I added the `UNKNOWN` (default) category to EFIS to make sure nothing changes for other formats, and the UI for changing the category type is still missing. So you could only create non-empty FF checklists starting from an existing FF checklist. I noted this in the PR, but since I'm more on the verbose side of things, that mention got lost in the stream of text.

It seems that it's okay for you to infest your model with FF group types, so I think `UNKNOWN` doesn't add any value under these circumstances, and tends to cause confusion like we just witnessed. In the new implementation, this should "just work" (tm)

rdamazio commented:
> > Original file: N425RP.json
> 
> I can't believe it, but apparently it wasn't possible to import the native format before, so I implemented it in [d6f7b58](https://github.com/rdamazio/efis-editor/commit/d6f7b58b69e0e1fdd88ae0184c80dffd0f6cae76). I hope you don't mind.

Good catch, thanks for adding that! I cherrypicked that change in since it's unrelated to Foreflight.

> > Hmmm something else is not right - I tried to export my plane's checklists as FMD with this code, and it gave me a 560-byte file which was empty :/
> 
> This was expected with my original implementation which decoupled EFIS categories from ForeFlight groups. I added the `UNKNOWN` (default) category to EFIS to make sure nothing changes for other formats, and the UI for changing the category type is still missing. So you could only create non-empty FF checklists starting from an existing FF checklist. I noted this in the PR, but since I'm more on the verbose side of things, that mention got lost in the stream of text.
> 
> It seems that it's okay for you to infest your model with FF group types, so I think `UNKNOWN` doesn't add any value under these circumstances, and tends to cause confusion like we just witnessed. In the new implementation, this should "just work" (tm)

Thanks. I tested again with my checklist and it looks like it's exporting properly now! :) I like the dropdown for selecting group type.

There was one bug I found - I have this checklist section:

<img width="1009" alt="Screenshot 2024-08-23 at 16 06 25" src="https://github.com/user-attachments/assets/4034eb92-59b0-43fd-9f4f-c03fb6e6e37b">

that got translated to:

```
...
{
  "objectId": "df137d3b4a674c7a80ed796dd845fbf9",
  "title": "O2 mode",
  "detail": "N(IGHT) OR D(AY)",
  "note": "(D)ay / (D)elayed mode turns on at 7,000ft MSL"
},
{
  "objectId": "8888b1738581482a92f3468574418a00",
  "title": "O2 auto-on (if desired)",
  "detail": "HOLD CLR/OFF TO TURN OFF, HOLD MODE FOR AUTO-ON",
  "note": "Auto-on turns on at 10,000ft"
},
...
```

so the "NOTE: (N)ight / (N)ow mode turns on immediately" entry got dropped completely. I guess consecutive notes like this could just be all joined together into the notes field? Need to see if Foreflight can deal with a multi-line note.

(but also feel free to finish this PR first and we can both do refinements like this afterwards, if you want)

rdamazio commented:
> > It seems that it's okay for you to infest your model with FF group types, so I think `UNKNOWN` doesn't add any value under these circumstances, and tends to cause confusion like we just witnessed. In the new implementation, this should "just work" (tm)
> 
> As the next logical step in making categories a first-class citizen in your internal representation, I've implemented not only display, but also editing in [070d938](https://github.com/rdamazio/efis-editor/commit/070d938ed80c11ecf9c15bd3e37d070f57904811). Not super proud of the code, but surprisingly it looks and works OK. I'd be interested in your thoughts on this:
> 
> Do you want to use the categories for other formats? I assume they don't support that. Are you otherwise okay with them getting lost on export for formats that can't display them?

Yes I'm fine with that - my principle in general is that no text should ever get dropped (since that'd mean someone may end up skipping an important checklist item), but that formatting/metadata can be dropped to adapt to the various formats. Does that sound reasonable?

zyv commented:
> Good catch, thanks for adding that! I cherrypicked that change in since it's unrelated to Foreflight.

Thanks, it will then be gone when I do the final rebase!

> There was one bug I found - I have this checklist section

Well, sorry, that was a kind of known artifact of not supporting multiline notes, and those in turn couldn't be supported until I mapped all items to ForeFlight concepts. This should be fixed now.

I've done some cursory testing, and your checklist generally looks good, but there are some ForeFlight specifics to consider. There's a lot of space between headings, so extra spacing isn't really needed. And it has a tendency to collapse long notes, so you will have to adjust according to your desired display device.

<img src="https://github.com/user-attachments/assets/43905e54-fb92-4ff8-8907-b0280d310b0b" width="76%">
<img src="https://github.com/user-attachments/assets/62592ddc-7cc9-431a-8494-95d744edbbbe" width="23%">

> Yes I'm fine with that - my principle in general is that no text should ever get dropped (since that'd mean someone may end up skipping an important checklist item), but that formatting/metadata can be dropped to adapt to the various formats. Does that sound reasonable?

Absolutely! I've now tried to map everything that makes sense to the ForeFlight concepts, and the only things that can get lost now should be indentation and centering. The rest should be reliably recoverable. I've tried to come up with a checklist for what's left, and I think we're getting closer! If you don't want to do the categories for other formats, I'll cross them out.

rdamazio commented:
> If you don't want to do the categories for other formats, I'll cross them out

It's not obvious to me how we would, so let's start without that?
(some formats don't even support groups, and I'm already taking liberties to adapt to that by only adding a group name to the checklist name after the first group - I don't think that adding the category to the checklist title for those would help at all).

rdamazio commented on src/model/proto/checklist.proto:
```diff
@@ -26,11 +26,17 @@ message ChecklistGroup {
-    NORMAL = 1;
-    ABNORMAL = 2;
-    EMERGENCY = 3;
+    normal = 0;
```
Totally optional, but FYI now that I saw your commit changing the JSON output for this: the reason for the UNKNOWN or UNSPECIFIED value is so that if you forget it, it doesn't look the same as if it's set to the first value (since in proto3 there's no required vs optional fields like in proto2).
https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum
(internally we actually prevent proto changes that don't follow this, but I think in this case it's pretty harmless if it's set to "normal" accidentally)

zyv replied:
This is indeed how I started for the reasons you mention, but then decided to back out and go against the rules to make it easier and safer(?). I implemented a conforming version in df2a24d. This has the consequence that all loading should be routed through the format class (which I guess is a good thing!), so I can explicitly make `normal` the default.

Otherwise I can add proper support for `unknown`, including a selection in the UI, but this will inevitably lead to questions like "why is my exported ForeFlight checklist empty? Or go back to breaking the rules.

I'm not sure what's best, I can live with both options, but I would be unhappy to implement `unknown` as a true default.

rdamazio replied:
I think what you did is right - unknown should be a synonym for "programming error" in this case, not something the user should ever have to deal with.

rdamazio commented on src/model/formats/test-data.ts:
```diff
@@ -178,12 +178,33 @@ export const EXPECTED_FOREFLIGHT_CONTENTS = {
+            {
+              prompt: '',
+              expectation: '',
+              type: ChecklistItem_Type.ITEM_PLAINTEXT,
```
ITEM_BLANK?

zyv replied:
> Should this be ITEM_BLANK?

Well, no, this is actually part of my specially designed evil correctness tests.

I split (and later collapse) multiline notes into indented text items. So if there is a completely blank line in a note, it should be translated into an indented text item, not a space. I use space items to denote empty space between items that cannot be checked off like checklist items in ForeFlight.

I'm afraid relaxing this will make the translation much more complicated and not necessarily more obvious to the user. At least nothing will get lost if you do it the "wrong" way, just the formatting will not be what you expect.

rdamazio commented on src/model/formats/test-data.ts:
```diff
@@ -178,12 +178,33 @@ export const EXPECTED_FOREFLIGHT_CONTENTS = {
               indent: 1,
               centered: false,
             },
+            {
```
Do you need to update other format test files for these? (or is that what you're asking me to do?)

rdamazio commented on src/model/formats/foreflight-reader.ts:
```diff
@@ -80,31 +80,53 @@ export class ForeFlightReader {
-          })
-        : checklistItem.detail
-          ? ChecklistItem.create({
+        ? // Detail Item
```
Style: these ternaries are getting pretty hard to follow, mind making them regular if/else control blocks?

zyv replied:
No problem, I've tried to improve the readability for you in 454463d - hopefully this makes it better.

I liked ternaries in this case because using a conditional expression instead of statements guarantees that you have considered all possible combinations of conditions. This is not synactically guaranteed with if-statements. But of course stronger correctness guarantees are of no use if the code is not understandable for humans :)

rdamazio commented on src/model/formats/foreflight-utils.ts:
```diff
@@ -22,6 +28,34 @@ export class ForeFlightUtils {
+
+  public static readonly CHECKLIST_ITEM_TYPES = new Map(
+    [...ForeFlightUtils.CHECKLIST_ITEM_PREFIXES.entries()].map(
+      (item) => item.reverse() as [string, ChecklistItem_Type],
```
Minor: The "as" here should not be needed if you do `new Map<ChecklistItem_Type, string>` and `new Map<string, ChecklistItem_Type>` above.
(Typescript tries to be strongly typed, so avoiding casting like this is a good idea)

zyv replied:
Thanks for the nudge, I didn't like the explicit cast either, but TypeScript's inference engine is just too stupid. The problem is that it couldn't infer that `reverse`ing a tuple would produce a tuple with swapped elements. So your suggestion would unfortunately not work.

However, it seems I found a solution in a2633eb - I implemented a Haskell-style `Data.Tuple` generic `swap` function, and now TypeScript is able to infer the types correctly. This has to do with the fact that tuples are not really tuples at all, but rather arrays, sigh.

rdamazio commented on src/model/formats/foreflight-utils.ts:
```diff
@@ -22,6 +28,34 @@ export class ForeFlightUtils {
+  );
+
+  public static promptToPartialChecklistItem(prompt: string): PartialChecklistItem {
+    const matches = /^([^:]+: )?(.*)$/.exec(prompt || '');
```
Do you need the first group to be consumed non-greedly?
i.e. "Foo: bar: baz" should probably get a prefix of "Foo: ", but with this it gets "Foo: bar: "

zyv replied:
> "Foo: bar: baz" should probably get a prefix of "Foo: ", but with this it gets "Foo: bar: "

Are you sure about this? This is exactly what I was trying to avoid with my `[^:]` hat trick. According to the regex tester and my unit test, it seems to work.

<img width="1189" alt="Screenshot 2024-08-26 at 11 38 22" src="https://github.com/user-attachments/assets/61050ac6-d8dd-45e8-84d2-e1840f514ab7">

rdamazio replied:
And that's what happens when you review code late at night :) You're right, sorry about that.

zyv replied:
Late night or not, sometimes one can overlook something... but always better to ask, could have been a real problem, so thank you anyway!

rdamazio commented:
Only very minor comments left - so looks like this is almost ready to merge? :) (am I forgetting anything that's still left?)
Once you think it's ready please rebase to the latest main (looks like there's some trivial conflicts to resolve) and let me know.

zyv commented:
> It's not obvious to me how we would, so let's start without that?

Agreed!

> Only very minor comments left - so looks like this is almost ready to merge? :)

Looks pretty good to me now, let me know if the latest changes are ok and I will do a rebase!

rdamazio commented:
> let me know if the latest changes are ok and I will do a rebase

Go for it!
Thanks again for this large contribution.

zyv commented:
> > let me know if the latest changes are ok and I will do a rebase
> 
> Go for it! Thanks again for this large contribution.

The pleasure is all mine! 👍
